src/v/raft/offset_translator.cc

```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "raft/offset_translator.h"
+
+#include "model/adl_serde.h"
+#include "raft/consensus_utils.h"
+#include "reflection/adl.h"
+#include "serde/serde.h"
+#include "storage/api.h"
+#include "vlog.h"
+
+namespace raft {
+
+static ss::logger logger{"offset_translator"};
+
+offset_translator::offset_translator(
+  std::vector<model::record_batch_type> filtered_types,
+  raft::group_id group,
+  model::ntp ntp,
+  storage::api& storage_api)
+  : _filtered_types(std::move(filtered_types))
+  , _group(group)
+  , _ntp(ntp)
+  , _logger(logger, ssx::sformat("ntp: {}", _ntp))
+  , _storage_api(storage_api) {}
+
+int64_t offset_translator::delta(model::offset o) const {
+    if (_filtered_types.empty()) {
+        return 0;
+    }
+
+    auto it = _offset2delta.lower_bound(o);
+    if (it == _offset2delta.begin()) {
+        throw std::runtime_error{_logger.format(
```

Comment:
I am wondering what about the cases where we need to translate log `start_offset` and first batch is the filtered one ?, should we make it consistent and return `_offset2delta.begin()->second - 1`

Reply:
The invariant that I try to uphold is that we always have an `_offset2delta` entry with offset before the log start offset. So if we translate something with offset `<= _offset2delta.begin()->first`, that's fishy. Any particular scenario that you have in mind?

Reply:
if there is always an entry with offset smaller than log start offset, it is perfect.